### DISCUSSION TITLE
Handling Slow Initial Page Load with Multiple Server Components and API Calls

### DISCUSSION THREAD

**Discussion by huxxnain** [vercel/next.js#73368] [Help]
### Summary

 I have a page with around 20 server components, and each component makes 2-3 API requests during the initial render. While the UI renders fine on page load, I’m facing a significant delay when interacting with the page for the first time.

For example, clicking a button that triggers an async action takes 3-4 minutes to execute. After this initial delay, everything works smoothly, but the issue is noticeable only on the first page load. It feels like the initial API requests block the thread, causing this delay.

Here’s what I’ve tried so far:

Implemented Suspense to manage async data fetching.
Experimented with different caching strategies.
Optimized individual components, but the issue persists.
Is there a way to avoid this blocking behavior on the first load? How can I ensure async actions work smoothly while the page loads data?

### Additional information

```js
const ProductLoader = async ({
  productId,
}: props) => {
  const response = fetch('/products') 

  if (response.state === 'error') {
    return <div>fail</div>
  }

 

  return (
    <Product
  
    />
  )
}

export { ProductLoader }
```

**Comment by denis23x** [comment]
Hi @huxxnain 

This is a lot of requests , try to combine/merge some eidpoints into one endpoint ..etc. Reduce the number of requests. Don't forget that the client also makes many requests for css, js, fonts and other.

As an option, you can run your requests in service workers.

https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers

**Comment by icyJoseph** [comment]
Are you doing waterfalls? If requests are independent of each other, make them parallel, with Promise.allSettled, and strip the results from there. 

Perhaps use the `use` hook, to stream data to the client, while other parts of the UI are loaded, https://react.dev/reference/react/use#streaming-data-from-server-to-client
